src/Sentry/Internal/MemoryMonitor.cs

```diff
@@ -0,0 +1,164 @@
+                       ?? throw new ArgumentException("No heap dump trigger configured on the options", nameof(options));
+        _onDumpCollected = onDumpCollected;
+        _onCaptureDump = onCaptureDump ?? CaptureMemoryDump;
+
+        _totalMemory = GC.GetGCMemoryInfo().TotalAvailableMemoryBytes;
+
+        GarbageCollectionMonitor.Start(CheckMemoryUsage, _cancellationTokenSource.Token);
+    }
+
+    internal void CheckMemoryUsage()
+    {
+        var eventTime = DateTimeOffset.UtcNow;
+        if (!_options.HeapDumpDebouncer.CanProcess(eventTime))
+        {
+            return;
+        }
+
+        var usedMemory = Environment.WorkingSet;
+        if (!_dumpTrigger(usedMemory, _totalMemory))
+        {
+            return;
+        }
+
+        _options.HeapDumpDebouncer.RecordOccurence(eventTime);
+
+        var usedMemoryPercentage = ((double)usedMemory / _totalMemory) * 100;
+        _options.LogDebug("Total Memory: {0:N0} bytes", _totalMemory);
+        _options.LogDebug("Memory used: {0:N0} bytes ({1:N2}%)", usedMemory, usedMemoryPercentage);
+        _options.LogDebug("Automatic heap dump triggered");
+        _onCaptureDump();
+    }
+
+    internal void CaptureMemoryDump()
+    {
+        if (_options.DisableFileWrite)
+        {
+            _options.LogDebug("File write has been disabled via the options. Unable to create memory dump.");
+            return;
+        }
+
+        var dumpFile = TryGetDumpLocation();
+        if (dumpFile is null)
+        {
+            return;
+        }
+
+        var processId = Environment.ProcessId;
+        _options.LogInfo("Creating a memory dump for Process ID: {0}", processId);
+
+        // Check which patth to use for dotnet-gcdump. If it's been bundled with the application, it will be available
+        // in the `dotnet-gcdump` folder of the application directory. Otherwise we assume it has been installed globally.
+        var bundledToolPath = Path.Combine(AppContext.BaseDirectory, "dotnet-gcdump", "dotnet-gcdump.dll");
+        if (File.Exists(bundledToolPath))
+        {
+            _options.LogDebug($"Using bundled version of dotnet-gcdump from: {bundledToolPath}");
+        }
+        else
+        {
+            _options.LogDebug("Using global version of dotnet-gcdump");
+        }
+
+        var arguments = $"collect -p {processId} -o '{dumpFile}'";
+        var command = File.Exists(bundledToolPath)
+            ? $"dotnet {bundledToolPath} {arguments}"
+            : $"dotnet-gcdump {arguments}";
+        var startInfo = new ProcessStartInfo
+        {
+            FileName = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "cmd.exe" : "/bin/bash",
+            Arguments = $"-c \"{command}\"",
+            RedirectStandardOutput = true,
+            UseShellExecute = false,
+            CreateNoWindow = true,
+        };
+
+        var process = new Process { StartInfo = startInfo };
+        process.Start();
+        while (!process.StandardOutput.EndOfStream)
```

Comment:
Do we need to subscribe also to stderr in addition to stdout?

Reply:
We could, if we wanted to log/format errors differently... in that case we could set `ProcessStartInfo.RedirectStandardError = true` and we'd get errors in a separate stream. 

Reading output from multiple streams comes with quite a bit of extra complexity though (reading [the docs](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.redirectstandarderror?view=net-8.0#remarks))... and probably not worth it in this case. If there are any problems, we'll detect these below, because the dump file won't be created correctly... and the reason why will be in the log just prior to logging that error (assuming Debug level diagnostic logging is enabled).